docs/examples/basic-price-oracle/main.go

```diff
@@ -0,0 +1,163 @@
+// This example demonstrates a basic RPC price oracle server that implements the
+// QueryRateTick RPC method. The server listens on localhost:8095 and returns a
+// rate tick for a given transaction type, subject asset, and payment asset. The
+// rate tick is the exchange rate between the subject asset and the payment
+// asset.
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	oraclerpc "github.com/lightninglabs/taproot-assets/taprpc/priceoraclerpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	// serviceListenAddress is the listening address of the service.
+	serviceListenAddress = "localhost:8095"
+)
+
+// RpcPriceOracleServer is a basic example RPC price oracle server.
+type RpcPriceOracleServer struct {
+	oraclerpc.UnimplementedPriceOracleServer
+}
+
+// isSupportedSubjectAsset returns true if the given subject asset is supported
+// by the price oracle, and false otherwise.
+func isSupportedSubjectAsset(subjectAsset *oraclerpc.AssetSpecifier) bool {
+	// Ensure that the subject asset is set.
+	if subjectAsset == nil {
+		return false
+	}
+
+	// In this example we'll only support a single asset.
+	assetIdStr := subjectAsset.GetAssetIdStr()
+	supportedAssetId := "7b4336d33b019df9438e586f83c587ca00fa65602497b9" +
+		"3ace193e9ce53b1a67"
+
+	return assetIdStr == supportedAssetId
+}
+
+// getRateTick returns a rate tick for a given transaction type and subject
+// asset max amount.
+func getRateTick(transactionType oraclerpc.TransactionType,
+	subjectAssetMaxAmount uint64) oraclerpc.RateTick {
+
+	// Determine the rate based on the transaction type.
+	var rate uint64
+	if transactionType == oraclerpc.TransactionType_PURCHASE {
+		// The rate for a purchase transaction is 42,000 milli-satoshi
+		// per asset unit
+		rate = 42_000
+	} else {
+		// The rate for a sale transaction is 40,000 milli-satoshi per
+		// asset unit
+		rate = 40_000
+	}
+
+	// Set the rate expiry to 5 minutes by default.
+	expiry := time.Now().Add(5 * time.Minute).Unix()
+
+	// If the subject asset max amount is greater than 100,000, set the rate
+	// expiry to 1 minute.
+	if subjectAssetMaxAmount > 100_000 {
+		expiry = time.Now().Add(1 * time.Minute).Unix()
+	}
+
+	return oraclerpc.RateTick{
+		Rate:            rate,
+		ExpiryTimestamp: uint64(expiry),
+	}
+}
+
+// QueryRateTick queries the rate tick for a given transaction type, subject
+// asset, and payment asset. The rate tick is the exchange rate between the
+// subject asset and the payment asset.
+func (p *RpcPriceOracleServer) QueryRateTick(_ context.Context,
+	req *oraclerpc.QueryRateTickRequest) (
+	*oraclerpc.QueryRateTickResponse, error) {
+
+	// Ensure that the payment asset is BTC. We only support BTC as the
+	// payment asset in this example.
+	if !oraclerpc.IsAssetBtc(req.PaymentAsset) {
```

Comment:
Is my understanding correct that `PaymentAsset`is the asset that'll be sent _outgoing_? Though if I think about the send vs recv flows, that means in the send flow the `PaymentAsset` is BTC, but in the recv flow (receiving some other asset) it'll be that asset? 

Or is it that this is the asset the penultimate hop will be receiving?

Reply:
I conceptualize the price oracle service as follows: it does not know on whether an asset is inbound or outbound on a channel; its primary function is to return a price. This applies regardless of whether the node is attempting to formulate a rate tick hint price or a quote accept rate tick.

Hence, it is the client's responsibility to select the `SubjectAsset` and `PaymentAsset` according to their needs. The `TransactionType` determines the trade side the client is interested in.

The price oracle is currently called in two places:
https://github.com/lightninglabs/taproot-assets/blob/eeca76c074bbc91197cfb67fb03cf384cd008b49/rfq/oracle.go#L310-L326

https://github.com/lightninglabs/taproot-assets/blob/eeca76c074bbc91197cfb67fb03cf384cd008b49/rfq/oracle.go#L228-L257

At both call sites the payment asset is BTC and it's the transaction type that varies.

I've tried to develop a basic price oracle in a generalized manner, such that it returns a price without any awareness of asset channels. I think that makes sense for light tap clients but maybe not for edge nodes?